Core changes to support running Splinter with allocated shared memory.

gapisback opened the pull request:
Support to run SplinterDB with shared memory configured for most memory allocation is an -EXPERIMENTAL- feature added with this commit.

This commit brings in basic support to create a shared memory segment and to redirect all memory allocation primitives to shared memory. Currently, we only support a simplistic memory mgmt; i.e. only-allocs, and no frees. With shared segments of 1-2 GiB we can run many functional and unit tests.

The high-points of the changes are:

- External configuration: splinterdb_config{} gains a few new visible fields to configure and troubleshoot shared memory configuration.
   - Boolean: use_shmem: Default is OFF
   - size_t : shmem_size:
   - Few boolean trace options to trace allocs / free from shmem.

- The main driving change is the re-deployment of platform_heap_id 'hid' arg that appears in all memory-related interfaces. If Splinter is configured for shared memory use, 'hid' will be an opaque handle to the shared segment. Most memory allocation will be redirected to new shmem-based alloc() / free() interfaces.

  The default 'hid' remains NULL, labelled as PROCESS_PRIVATE_HEAP_ID.

- Manage handling of heap-ID and heap-handle to platform_get_heap_id() to correctly return the handle to shared memory. (Otherwise, it would return NULL by default.)

- Formalize usages of PROCESS_PRIVATE_HEAP_ID: A small number of clients that wish to repeatedly allocate large chunks of memory tend to cause OOMs. The memory allocated by these clients is not shared across threads / processes. For such usages, introduce PROCESS_PRIVATE_HEAP_ID as an alias to NULL. The known set of clients using this interface now are:

    - merge_iterator_create(), merge_iterator_destroy()
    - trunk_range() - trunk_range_iterator_init(), trunk_range_iterator_deinit()
    - routing_filter_add() - temp array

- BTree pack allocates large fingerprint-array. This also causes large tests to run into OOMs. For threaded execution, it's ok if the memory for this array is allocated from the heap. But for multi-process execution, when one process (thread) allocates this finger print array, another thread may pick up the task to compact a bundle and will try to free this memory. So, this memory has to come from shared memory. To cope with such repeated allocations of large chunks of memory to build fingerprint, a small recycling scheme for "free"-large-memory chunks scheme is now supported by shmem module.

  Use this technique to recycle memory allocated for iterators also. They tend to be big'gish, so can also cause shmem-OOMs.

- Several existing functional and unit-tests have been enhanced to now support "--use-shmem" as the 1st option. This will create Splinter with shared memory configured, and tests run in this mode. This change brings-in quite a good coverage of existing testing for this new feature. Following tests are working with this feature:
     - limitations_test - splinterdb_quick_test - writable_buffer_test - task_system_test - splinter_test - btree_test - splinterdb_stress_test

   - New test: large_inserts_bugs_stress_test -- added to cover the primary use-case of concurrent insert performance benchmarking (that this feature is driving in prior integration effort).

   - test.sh enhanced to run different classes of test with "--use-shmem" option.

- Diagnostis & Troubleshooting:

   - Shmem-based alloc/free interfaces extended to print name of object and other call-site info, to better pinpoint source code-flow leading to memory issues.

   - Add shared memory usage metrics, including for large-fragment handling. Report summary-line of metrics when Splinter is shutdown. Print stats on close.

   - Add various utility diagnostic helper methods to validate that addresses within shared memory are valid. Unit-tests and some asserts use these.

gapisback commented on include/splinterdb/splinterdb.h:
```diff
@@ -32,12 +53,18 @@ typedef struct {
    void *heap_id;
 
+   // Shared memory support
+   uint64 shmem_size;
```
Few externally visible config options. Intended for internal use now. See big block comment above.

gapisback commented on include/splinterdb/splinterdb.h:
```diff
@@ -32,12 +53,18 @@ typedef struct {
+   _Bool  use_shmem; // Default is FALSE.
+   _Bool  trace_shmem_allocs;
+   _Bool  trace_shmem_frees;
+   _Bool  trace_shmem; // Trace both allocs & frees from shared memory
```
Ideally, we should have one `_Bool trace_flags` bitmap, where each feature needing tracing turns ON its bit.

I didn't go that far, yet, ... unsure how the team feels about this extra tracing baggage.
I did find these traceflags very useful during dev & stabilization, so would like to retain these.

gapisback replied:
Re-discussed with @rtjohnso -- 

Make these global bool variables in shmem.c and take them out of this public interface `struct splinterdb_config{}`

In config_parse(), when the `--trace` flag is supplied, toggle these boolean variables.

gapisback commented on include/splinterdb/splinterdb.h:
```diff
@@ -136,7 +162,7 @@ typedef struct splinterdb splinterdb;
 // So it must live at least as long as the splinterdb
 int
-splinterdb_create(const splinterdb_config *cfg, splinterdb **kvs);
+splinterdb_create(splinterdb_config *cfg, splinterdb **kvs);
```
See the definition for the reasons why changed from `const` to non-const. Same applies to `splinterdb_open()` below.

gapisback replied:
Has been fixed.

gapisback commented on src/btree.h:
```diff
@@ -342,8 +342,18 @@ btree_pack_req_init(btree_pack_req  *req,
+                   "Unable to allocate memory for %lu tuples",
+                   max_tuples);
+      if (!req->fingerprint_arr) {
+         return STATUS_NO_MEMORY;
```
Added OOM error handling percolation. Hardening ...

gapisback commented on src/clockcache.c:
```diff
@@ -915,12 +920,20 @@ clockcache_assert_no_locks_held(clockcache *cc)
+                   " status. Found unexpected status=0x%x %s\n",
+                   i,
+                   entry_flag,
+                   ((entry_flag & CC_ACCESSED) ? "(CC_ACCESSED)" : ""));
```
Expanded debug asserts ... Was running into this before I figured out the need for the fix to `mmap()`. 

Leave it.... will benefit some future engineer.

gapisback commented on src/data_internal.c:
```diff
@@ -41,7 +41,8 @@ merge_accumulator_copy_message(merge_accumulator *ma, message msg)
 {
-   platform_status rc = writable_buffer_resize(&ma->data, newsize);
+   platform_status rc =
+      writable_buffer_resize(&ma->data, merge_accumulator_length(ma), newsize);
```
I don't recall why / what this change is for.  Don't even recall if I added it.
Seems like an inherit from `/main` ?

@rtjohnso -- we should re-visit during review and finalize.

gapisback commented on src/merge.c:
```diff
@@ -421,7 +421,7 @@ merge_iterator_create(platform_heap_id hid,
                   "size mismatch");
 
-   merge_itor = TYPED_ZALLOC(hid, merge_itor);
+   merge_itor = TYPED_ZALLOC(PROCESS_PRIVATE_HEAP_ID, merge_itor);
```
Renamed to this mnemonic (rather than `NULL_HEAP_ID`, which is what I had used during proto).

This is inline with a request from Rob for a better name.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -19,6 +19,15 @@ bool platform_use_mlock   = FALSE;
+ * are in use.
+ */
+platform_heap_handle Heap_handle = NULL;
+platform_heap_id     Heap_id     = NULL;
```
`Heap_id` is needed to service the return from `platform_get_heap_id()`.

The sibling global is created for symmetry.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -47,21 +56,42 @@ platform_get_stdout_stream(void)
                      platform_heap_id     *heap_id)
 {
+   if (use_shmem) {
+      platform_status rc = platform_shmcreate(max, heap_handle, heap_id);
```
Here is where the Magik! begins.

gapisback commented on src/platform_linux/platform_inline.h:
```diff
@@ -34,11 +36,13 @@ platform_checksum_is_equal(checksum128 left, checksum128 right)
-   free(ptr);
-}
+static void
+platform_free_from_heap(platform_heap_id UNUSED_PARAM(heap_id),
```
`heap_id` is really used in the new version, so this should go away.

Surprisingly -- no compiler complained about this inconsistency. (I'll let this be, and will fix this in part-3.)

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+                   "Heap handle is expected to be NULL while creating a new "
+                   "shared segment.\n");
+
+   int shmid = shmget(0, size, (IPC_CREAT | PLATFORM_IPC_OBJS_PERMS));
```
@rosenhouse wanted me to implement this with System-V IPC APIs. I didn't get around to that.

I'd like this to stabilize and integrate. Will re-visit this aspect in part-3 PR while reviews for part-1 and part-2 are on-going. I don't think this is a big blocker issue right now. 

Let me know your opinions.

gapisback replied:
@rtjohnso & I re-reviewed this in closing meeting. For now, as this functionality works, and it allows us someday to be able to do `shmattach()` -- we can live with this.

Nothing further to do at this time.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+ * the heap-ID handle to the shared memory's start address, from which the
+ * location of the next-free-byte can be tracked.
+ */
+static inline platform_heap_handle
```
WHOLE bunch of conversion / access lookup functions. You can skip this pile.

gapisback replied:
Deleted `platform_heap_handle *` from the code base and from this file. Reworked mapping interfaces to only use `platform_heap_id`.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+      // Copy over old contents, if any, and free that memory piece
+      if (oldptr) {
+         memcpy(retptr, oldptr, oldsize);
+         splinter_shm_free(hid, oldptr, "Unknown");
```
This `realloc()` logic ... is working in this part-1. I had to overhaul and correct this implementation in part-3, after the pull from `/main` for Rob's large-values support. For now, let this be as-is; it's a working version.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   // See if we can satisfy requests for large memory fragments from a cached
+   // list of used/free fragments that are tracked separately.
+   if ((size >= SHM_LARGE_FRAG_SIZE)
+       && ((retptr = platform_shm_find_free(
```
Control farms-off here to handle large free-fragment recycling scheme. Look into this minion later on.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   __sync_fetch_and_sub(&shminfo->shm_free_bytes, size);
+
+   if (size >= SHM_LARGE_FRAG_SIZE) {
+      platform_shm_track_alloc(shminfo, retptr, size);
```
Large fragment allocation is tracked here. Imp ...

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+      return;
+   }
+
+   platform_shm_track_free(shminfo, ptr, objname, func, file, lineno);
```
Freeing large fragments is tracked here, for future recycling.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   // large fragments tracked, then this is certainly not a large
+   // fragment. So, no further need to see if it's a tracked fragment.
+   if (addr > shm->shm_large_frag_hip) {
+      shm->shm_nf_search_skipped++; // Track # of optimizations done
```
This is an important optimization to avoid searching needlessly in loop below when small fragments are freed.

Recall that in this version we don't have a real free().

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+      // Mark the fragment as in-use by recording the process/thread that's
+      // doing the free.
+      frag->shm_frag_freed_by_pid = getpid();
+      frag->shm_frag_freed_by_tid = platform_get_tid();
```
All this `debug_assert` and freed_by_pid / allocated_to_pid business has worked very well.

With this all in-place from the get-go, I didn't run into a single bad code / erroneous large-free-fragment ownership in any of the high-volume benchmarking.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+ * context fields, which are printed for diagnostics under a traceflag.
+ */
+#define splinter_shm_free(heap_id, ptr, objname)                               \
+   platform_shm_free(heap_id, ptr, objname, func, file, lineno)
```
Ignore this kind of wrapper caller-macros that you will see in part-1 and part-2.

I've streamlined the interfaces and call-sequences in part-3. We can review all that later and finalie.

gapisback commented on src/splinterdb.c:
```diff
@@ -137,8 +142,8 @@ splinterdb_validate_app_data_config(const data_config *cfg)
 static platform_status
-splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
-                       splinterdb              *kvs      // OUT
+splinterdb_init_config(splinterdb_config *kvs_cfg, // IN
```
This change of taking out `const` qualifier can come out. I had it like-this in a prior dev version, but now I see that in this final version, it's only read-only. Will change this back.

gapisback replied:
Revisit & fix this ... this came up on further discussions with Rob.

gapisback replied:
Resolved. Reworked interfaces to re-introduce `const splinterdb_config *kvs_cfg`.

Cleaned up backout code. Added new unit test case.

gapisback commented on src/splinterdb.c:
```diff
@@ -227,38 +229,95 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
-splinterdb_create_or_open(const splinterdb_config *kvs_cfg,      // IN
-                          splinterdb             **kvs_out,      // OUT
-                          bool                     open_existing // IN
+splinterdb_create_or_open(splinterdb_config *kvs_cfg,      // IN
```
The reason why `kvs_cfg` is losing the `const` attribute is due to the change(s) below on L310 & 311. See reasons below.

gapisback replied:
Resolved. Reworked interfaces to re-introduce const splinterdb_config *kvs_cfg.

gapisback commented on src/splinterdb.c:
```diff
@@ -227,38 +229,95 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
+   // only create one if it does not already exist.)
+   if (kvs_cfg->use_shmem && (kvs_cfg->heap_handle == NULL)) {
+      size_t shmem_size = (kvs_cfg->shmem_size ? kvs_cfg->shmem_size : 2 * GiB);
+      status            = platform_heap_create(platform_get_module_id(),
```
Attn: @rtjohnso - In Oct or so of 2022, when I first aired out the proto-version PR of this work, you had some major concern about this block of code. I didn't register it then as that airing was only a preview and wasn't quite ready for a formal review.

I forget now what your issue was about this chunk of code.

I am calling it out now, in case it jogs your memory as to what about this code piece you wanted a rework.

Let's discuss. ...

gapisback commented on src/splinterdb.c:
```diff
@@ -227,38 +229,95 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
+
+   // This allows for a simple usage where application can close and reopen
+   // Splinter and not run into seg-faults due to stale memory handles.
+   kvs_cfg->heap_handle = NULL;
```
This is why the input `const` qualifier was dropped. There is at least one test case that wants to close & reopen Splinter. If the heap-handles off of input `kvs_cfg` are retained, then due to the way the logic is implemented on L 247 above, for the usage scenario of a close-and-reopen we will end up with an invalid `kvs_cfg->heap_handle`.

All hell breaks loose.

This rework accommodates all use-cases. If there is a better way to rework this exercise / usage, I'll be happy to hear suggestions.

gapisback replied:
Resolved. Reworked interfaces to re-introduce const splinterdb_config *kvs_cfg.

gapisback commented on src/splinterdb.c:
```diff
@@ -338,23 +397,33 @@ splinterdb_create_or_open(const splinterdb_config *kvs_cfg,      // IN
+   // different place. Use one carefully, to avoid ASAN-errors.
+   platform_free((kvs->heap_id ? kvs->heap_id : kvs_cfg->heap_id), kvs);
+   if (we_created_heap) {
+      platform_heap_destroy(&heap_handle);
```
This whole error-handling backout block is touchy. I have finessed it "just-right" for all the existing test cases that go thru this 'wrong-param-fail-error-check' test scenarios.

gapisback replied:
Resolved. Reworked interfaces to re-introduce const splinterdb_config *kvs_cfg.

gapisback commented on src/task.c:
```diff
@@ -644,6 +650,7 @@ task_enqueue(task_system *ts,
    }
 
    __sync_fetch_and_add(&group->current_waiting_tasks, 1);
+   __sync_fetch_and_add(&ts->ntasks_enqueued, 1);
```
Metric to track how many tasks were enqueued.

This came very much in handy while running multi-threaded heavy insert workloads, to look closer into fg/bg-thread machinery is playing out.

gapisback commented on src/trunk.c:
```diff
@@ -7047,7 +7088,7 @@ trunk_range(trunk_handle  *spl,
 destroy_range_itor:
    trunk_range_iterator_deinit(range_itor);
-   platform_free(spl->heap_id, range_itor);
+   platform_free(PROCESS_PRIVATE_HEAP_ID, range_itor);
```
One of the instances where we fall-back to `malloc()` from the heap.

Once part-3 is finalized and integrated, it will be a good opportunity to clean this up. Now that we have free-fragment recycling working in part-3, this kind of bypass may probably be unnecessary. But I'd rather not have to do anything about it right now, till all this work gets integrated.

gapisback commented on src/util.c:
```diff
@@ -19,7 +21,7 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
 
    void *oldptr  = wb->can_free ? wb->buffer : NULL;
-   void *newdata = platform_realloc(wb->heap_id, oldptr, minspace);
+   void *newdata = platform_realloc(wb->heap_id, oldspace, oldptr, minspace);
```
The `oldspace` arg is needed for shm-based `realloc`.

gapisback commented on test.sh:
```diff
@@ -530,70 +566,138 @@ function test_make_run_tests() {
+   "$BINDIR"/unit/task_system_test $use_shmem --num-bg-threads 4 --num-memtable-bg-threads  2
+
+   # shellcheck disable=SC2086
+   "$BINDIR"/driver_test io_apis_test $use_shmem
```
All of these fast unit-tests can now be run with `--use-shmem` option.  You will get this coverage in every test run.

gapisback commented on test.sh:
```diff
@@ -530,70 +566,138 @@ function test_make_run_tests() {
+    run_with_timing "${msg}" \
+        "$BINDIR"/unit/large_inserts_bugs_stress_test ${use_shmem} \
+                                                      --num-threads ${n_threads} \
+                                                      --num-normal-bg-threads 4 \
```
Some slower-running unit-tests have also been enhanced to run with `--use-shmem` option.

Again, more testing coverage for shared-memory integration is ON by default in every CI run.

gapisback commented on test.sh:
```diff
@@ -530,70 +566,138 @@ function test_make_run_tests() {
+    # shellcheck disable=SC2086
+    run_with_timing "Functionality test, key size=maximum (${max_key_size} bytes)${use_msg}" \
         "$BINDIR"/driver_test splinter_test --functionality 1000000 100 \
+                                            $use_shmem \
```
Oh, I seem to have upgraded a few functionality and `--perf` tests, too, (see below) to run with `--use-shmem`.

Even more testing coverage is ON by default in CI!

gapisback commented on test.sh:
```diff
@@ -661,6 +769,25 @@ function run_other_driver_tests() {
+# may not still be runnable in this mode. So, we will incrementally online
+# remaining tests when they can run successfully in this mode.
+# #######################################################################
+function run_tests_with_shared_memory() {
```
OK, here's a **quick-tip**. 

I added a "feature" to `test.sh` in `/main` previously by which you can invoke each of these sub-functions. Here is the help output from `test.sh`:

```
To run a smaller collection of slow running tests,
name the function that drives the test execution.
Examples:
  INCLUDE_SLOW_TESTS=true ./test.sh run_btree_tests
  INCLUDE_SLOW_TESTS=true ./test.sh run_splinter_functionality_tests
  INCLUDE_SLOW_TESTS=true ./test.sh nightly_cache_perf_tests
```

So, if you want to specifically re-run just ALL those tests that exercise shared-memory support, you would do:

`$ INCLUDE_SLOW_TESTS=true ./test.sh run_tests_with_shared_memory`


Greatly improves dev productivity when stabilizing some new code you added.

gapisback commented on tests/config.c:
```diff
@@ -308,13 +332,45 @@ config_parse(master_config *cfg, const uint8 num_config, int argc, char *argv[])
+            }
+         }
+         config_set_mib("shmem-capacity", cfg, shmem_size) {}
+         config_set_gib("shmem-capacity", cfg, shmem_size) {}
```
some functional and unit-tests that use `config_parse()` interface and are enhanced to run with `--use-shmem` can be run with diff configurations of shared memory size.

Going forward, this will be specifically useful for functional `splinter_test` and also for perf tests.

gapisback commented on tests/unit/splinter_ipc_test.c:
```diff
@@ -0,0 +1,32 @@
+/*
+ * Basic test case.
+ */
+CTEST2(splinter_ipc, test_basic_flow) {}
```
This was an early attempt, and has been withdrawn eventually. Ignore this.

rtjohnso replied:
Ignore the whole file?

gapisback replied:
Deleted this file as no new tests were added to this file in future change sets.

gapisback commented on tests/unit/splinter_shmem_test.c:
```diff
@@ -0,0 +1,658 @@
+ * -----------------------------------------------------------------------------
+ * splinter_shmem_test.c --
+ *
+ *  Exercises the interfaces in SplinterDB shared memory allocation module.
```
Very useful and interesting shmem-focussed testing.

Has been very helpful to me during shmem-dev.

Test has been greatly enhanced and expanded, especially in part-3.

gapisback commented:
**Note to the reviewers**: @rtjohnso @ajhconway @rosenhouse :

The entire shared-memory support is being parceled into 3-part change-sets. This is the 1st of the whole work.

1. Core shared memory support
2. Support for multi-process, forked-processes executing with shared memory
3. Support for free-fragment recycling, improved shared memory management.

I am working on putting together the PRs for (2) and (3).

Here is the order to review files in, to get a good grip on this change-set:

1. platform.h, platform.c
4. shmem.h,  shmem.c: `platform_shmcreate()`, `platform_shmdestroy()`, `platform_shm_alloc()`, `platform_shm_free()`, 
5. Misc other header files
6. Review `test.sh` to gauge the scope of changed and enhanced testing brought-in with this work.

rtjohnso reviewed (changes requested):
Here are my reviews of the platform-level stuff.

Thanks, Aditya!

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+
+#pragma once
+
+#include <sys/types.h>
```
Update: 

- shmem.h can possibly be removed from trunk.h / .c
- Try to not to leak the shmem-extern APIs into common code.

gapisback replied:
Resolved. Updated code as discussed.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+#include <sys/types.h>
+#include <sys/shm.h>
+
+typedef struct shmem_info shmem_info;
```
Agreed to rename this to `typedef struct shmem_heap shmem_heap;`

gapisback replied:
Done. Removed L9 from the `.h` file and renamed struct to `struct shmem_heap`

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+ * Interfaces to retrieve size(s) using heap_handle.
+ */
+size_t
+platform_shmsize_by_hh(platform_heap_handle heap_handle);
```
Re-discussed this ... Rob still feels this can be deleted from `shmem.c / .h` code. I can leave heap_handle in the code base, and leave existing usages untouched. Just do not use it in new shmem code.

gapisback replied:
Resolved. Lots of code cleanup done. Have tried to remove most references / usages of `platform_heap_handle` in much of the code base. This removes / cleans-up areas where that parameter was being used in `shmem.c`.

Final purge of this `heap_handle *` from rest of splinter code ... can be done separately.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+typedef struct shmem_info shmem_info;
+
+platform_status
+platform_shmcreate(size_t                size,
```
Agreed to:

- Change to: `platform_shmcreate(size_t size, shmem_heap *heap_id);
- In caller, pass-in `shmem_heap **`, and callee returns output ptr.
- All other interfaces will want to change to something like `shmem_heap *`.

gapisback replied:
Done. Let's review on final discussion.

gapisback replied:
@rtjohnso - Update, after having gone thru most of the rework to cleanup and stabilize this change-set.

I see that I have not quite done what you requested, which is to change this interface to become:

`platform_shmcreate(size_t size, shmem_heap **heap_id)`

One reason is that with the rework done, the definition of `struct shmem_heap{}` is local to shmem.c, and the single external typedef refer in `shmem.h` has been taken out.

In order to do what you asked, the flow from `platform_heap_create()` -> `platform_shmcreate()`, would now have 'know' about `struct shmem_heap` in platform.c .

I didn't find that approach hugely better ... so stopped at what I did so far.

We can still re-think if you feel really strongly that the suggested rework must be done in this round.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+#include "util.h"
+
+// SplinterDB's shared segment magic identifier
+#define SPLINTERDB_SHMEM_MAGIC (uint64)0xDEFACADE
```
Discussed and we agreed to take it out.

gapisback replied:
@rtjohnso - I have reworked the interfaces in this file to consistently take-in `shmem_heap *` parameter.

I have also purged the `platform_heap_handle *` from this file.
All interfaces in this module now work with `platform_heap_id *hid` and `shmem_heap *`.

This brings the code closer to your wish of using the type-system to defend against bad shmem handle pointers.

But, even after this cleanup, and due to my own refactoring, the following check and error in `platform_shmdestroy()` was still tripping-up while running some unit-tests:

```
 334    if (shm->shm_magic != SPLINTERDB_SHMEM_MAGIC) {
 335       platform_error_log("%s(): Input heap ID, %p, does not seem to be a "
 336                          "valid SplinterDB shared segment's start address."
 337                          " Found magic 0x%lX does not match expected"
 338                          " magic 0x%lX.\n",
 339                          __func__,
```

I have, since, discovered and fixed this self-introduced bug in this line of code in the destroy function:

```
 320    const void *shmaddr = (const void *)platform_heap_id_to_shmaddr(*hid_out);
```

`platform_heap_id_to_shmaddr()` needed to be called by `*hid_out` and not `hid_out`.

The point being: Given that this check on SHMEM_MAGIC did come handy even during my own refactoring and stabilization, I see some value for retaining this (even with the improvements in various interfaces that now take typed arguments).

So, even thought I think you objected to this poor-man's consistency checking, I wish to retain it, so it will provided one more last-line-of-defense against code refactoring in the future.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+ */
+typedef struct shmem_info {
+   void *shm_start; // Points to start address of shared segment.
+   void *shm_end;   // Points to end address; one past end of sh segment
```
Discussed, and Rob relented ... and agreed to leave it in.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   uint32        shm_num_frags_inuse_HWM;
+   size_t        shm_used_by_large_frags; // Actually reserved
+   size_t        shm_nfrees;              // # of calls to free memory
+   size_t        shm_nf_search_skipped;   // See above;
```
Discussed... Rob prefers that this come out. Will do so ...

gapisback replied:
Took it out as suggested. (Upon a re-visit a few months since we spoke, I do think this kind of monitoring counter will be beneficial during perf benchmarking.)

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   *heap_handle = NULL;
+
+   // Use a cached copy in case we are dealing with bogus input shmem address.
+   shmem_info shmem_info_struct;
```
Discussed; will do as suggested.

gapisback replied:
@rtjohnso - I tried to rework this to take this validity detection out.

But, as explained in the section for `SPLINTERDB_SHMEM_MAGIC`, I did find this useful even in my own stabilization cycles. Retained.

rtjohnso commented on tests/functional/filter_test.c:
```diff
@@ -47,7 +47,7 @@ test_filter_basic(cache           *cc,
 
    DECLARE_AUTO_WRITABLE_BUFFER(keywb, hid);
-   writable_buffer_resize(&keywb, key_size);
+   writable_buffer_resize(&keywb, 0, key_size);
```
Same comment as always...

rtjohnso commented on tests/test_common.c:
```diff
@@ -8,6 +8,8 @@
  * Module contains functions shared between functional/ and unit/ test sources.
  * -----------------------------------------------------------------------------
  */
+#include <unistd.h>
```
Why?

gapisback replied:
don't recall why it was added. Took it out ... builds seem to work. Will re-test in CI.

rtjohnso commented on tests/test_misc_common.h:
```diff
@@ -0,0 +1,30 @@
+ * run the test(s) using shared memory segment setup & allocation.
+ */
+static inline bool
+test_using_shmem(int argc, char *argv[])
```
I'm confused.  Doesn't the generic config parsing already handle detecting that flag?

rtjohnso replied:
Oh this is for the unit tests.

How about just making them all run all tests both w/ and w/o shmem?  No command-line parameter, which is the model we've been following so far for unit tests.

gapisback replied:
@rtjohnso - This arg parsing and use of `config_parse()` v/s the newly added `test_using_shmem()` is a bit wonky. (I agree.)

config_parse does parse `--use-shmem`. But there is an ordering issue here in when that arg is parsed and when it has to be used.

A good example is the `tests/unit/btree_test.c` test. There we need to `platform_heap_create()` before the block of code that does other parsing and config-setup is done.

I _could_ rearrange this logic to have `config_parse()` do the required parsing. But that does not solve the problem entirely.

Take a look at `tests/unit/limitations_test.c`: There I just want to run this test with and w/o `--use-shmem` argument.

If I were to change that test to also now parse arguments using `config_parse()`, that will open up the gates to call this test with many more other parameters -- none of which are either meant for or tested-for in this test.

That's why I resorted to this `test_using_shmem()` approach, requiring that the `--use-shmem` arg be the very 1st arg to a unit-test, if you want to run it with shared memory enabled.

Reworking this entire argument parsing is quite a bit of work, and I really would not want to take it on now.

gapisback replied:
And about: "How about just making them all run all tests both w/ and w/o shmem?"

You cannot just run a unit-test in both modes. To do that, we'd have to rework the unit-test execution machinery, something around the main() of ctests to run each test in both modes.

But then, not all tests are enabled to run w/ shared-memory. That's why you have to invoke them with the right argument.

The logic to execute is managed by commands in `test.sh`.

I think this test execution environment / usage is something we should take on far after all 3 parts of this feature dev is integrated to /main.

If I try to clean this up now, it will distract me further and will delay moving the remaining 2 parts forward.

gapisback replied:
Discussed with @rtjohnso -- We agreed to leave this as is for now ... as other approaches will require quite a bit of updates to config_parse() and related machinery.

gapisback replied:
After another round of review examining this code, we agreed to withdraw this newly added `test_using_shmem()` interface.

Deleted this special config parsing fn, and deleted this file.

Reworked all tests that were calling this function to now call `config_parse()` instead, to check if the test is being run with `--use-shmem` argument.

rtjohnso commented on tests/functional/io_apis_test.c:
```diff
@@ -153,13 +159,24 @@ npages_per_thread(io_test_fn_args *io_test_param, int nthreads)
+   argv++;
+
+   // If --use-shmem was provided, parse past that argument.
+   bool use_shmem = (argc >= 1) && test_using_shmem(argc, (char **)argv);
```
Doesn't config_parse() handle that flag?

gapisback replied:
Yes, this can be changed. Slipped my scan thru review comments previously.

Reworked this now to use `config_parse()` as the interface to parse the `--use-shmem` arg.

rtjohnso commented on tests/functional/splinter_test.c:
```diff
@@ -2608,6 +2620,13 @@ splinter_test(int argc, char *argv[])
       config_argv += 1;
    }
+   if (config_argc > 0
+       && strncmp(config_argv[0], "--use-shmem", sizeof("--use-shmem")) == 0)
```
Doesn't the generic config parsing handle this?

gapisback replied:
@rtjohnso - Well, ... `config_parse()` does parse this `--use-shmem` arg, but that function is not being used in this test to parse cmdline args. This test has very painful parsing of fixed-positions of arguments. 

And somewhere along the line it's calling this even-further specialized `test_config_parse()` parsing function.

So it got difficult to use `config_parse()` to recognize use of `--use-shmem`.

rtjohnso commented on tests/unit/large_inserts_bugs_stress_test.c:
```diff
@@ -0,0 +1,661 @@
+// Copyright 2022 VMware, Inc.
```
Haven't we rejected adding this test in the past?  I don't see a lot of shmem-specific stuff to it.

gapisback replied:
@rtjohnso - Yes, we discussed this previously. Over the 3-part stabilization this test has added value in finding bugs that are otherwise not found by existing stress / perf tests.

Even in this round of stabilization, if I enable the calls to this test from `test.sh`, some cases will randomly fail. We surface trunk bugs and so on.

Stabilizing all those as part of this feature dev integration is way too much than I can take on now.

For now, I have left this file in the change set, and it will still be compiled and built.

As we move thru part-2 and part-3, you will see some merges for this file. 

I think we should revisit and reconsider whether to bring this test suite online after part-3 is completely reviewed and integrated.

Let's discuss ... in our closure meeting.

gapisback commented on src/platform_linux/platform_inline.h:
```diff
@@ -400,10 +405,31 @@ platform_get_module_id()
+ */
 static inline void *
-platform_aligned_malloc(const platform_heap_id UNUSED_PARAM(heap_id),
+platform_aligned_malloc(const platform_heap_id heap_id,
```
Done.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+static inline void *
+platform_shm_hip(platform_heap_id hid)
+{
+   return (((shmem_info *)platform_heap_id_to_shmaddr(hid))->shm_end - 1);
```
I re-looked at this `hip` and its semantic seems useful enough to me. `lop` and `hip` give you the valid range of min / max addresses from shared memory. There is only one usage right ... I'd like to retain it.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+                     const int        lineno);
+
+static inline int
+platform_shm_alignment()
```
right, it's unused. Deleted this interface.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   shm_frag_info *frag = shm->shm_mem_frags;
+   for (int fctr = 0; fctr < ARRAY_SIZE(shm->shm_mem_frags); fctr++, frag++) {
+      if (!frag->shm_frag_addr || (frag->shm_frag_addr != addr)) {
+         continue;
```
Hmmm ... not sure I understand this comment / ask. We are checking if a fragment-tracker is free (not tracking a free fragment) or if it's validly tracking a free fragment that we are interested in, given by `addr`.

If we do find this fragment's address; i.e. `frag->shm_frag_addr == addr`, then we need to do some book-keeping.

That's why the continue, and the rest of the code still needs to do something.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+      if (frag->shm_frag_size < size) {
+         continue;
+      }
+      found_tracked_frag = TRUE;
```
(Again -- not sure I follow your suggestion. I think the loop has to process one fragment, and there is a `break` at the end of the loop.

I could move that body outside the loop, but then will have to still check if `fctr` is valid.

Am not entirely convinced that your suggested rewrite actually simplifies the code-flow. (We can talk in person.)

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+ * hooks.
+ * -----------------------------------------------------------------------------
+ */
+bool
```
Deleted `platform_shm_heap_handle_valid()`

gapisback commented on src/trunk.h:
```diff
@@ -19,6 +19,7 @@
 #include "allocator.h"
 #include "log.h"
 #include "srq.h"
+#include "shmem.h"
```
(Seems like you made a sweep and fixed some `.h` inclusions.)

gapisback commented on tests/config.h:
```diff
@@ -9,7 +9,7 @@
 #pragma once
 
-#include "clockcache.h"
+#include "clockcache.h" // Nested #include's task.h and related tokens.
```
Yes, that's fine. Thank you for the cleanup.

gapisback commented on src/clockcache.c:
```diff
@@ -915,12 +920,20 @@ clockcache_assert_no_locks_held(clockcache *cc)
+
+      // We expect entry to be in only one of these two states.
+      entry_status entry_flag = clockcache_get_flag(cc, i);
+      debug_assert(((entry_flag & (CC_FREE | CC_CLEAN)) != 0),
```
I had rewritten it so that the assert-message could print out the `entry_flag` status value.

As per your suggestion, reinstated the `debug_assert()` the way it was previously.

The `#if SPLINTER_DEBUG` is just to make sure that this code is only active in debug-builds. The rest of the code, i.e. body of `for()` is an empty block as all it contains is the `debug_assert()`.

gapisback commented on src/routing_filter.c:
```diff
@@ -396,7 +396,8 @@ routing_filter_add(cache           *cc,
    debug_assert(temp_buffer_count < 100000000);
-   uint32 *temp = TYPED_ARRAY_ZALLOC(hid, temp, temp_buffer_count);
+   uint32 *temp =
+      TYPED_ARRAY_ZALLOC(PROCESS_PRIVATE_HEAP_ID, temp, temp_buffer_count);
```
`TYPED_ARRAY_ZALLOC` eventually snakes through `TYPED_MALLOC`, -> ... `platform_aligned_malloc()` -> `splinter_shm_alloc()`.

heap_id is used in the lower-most function, so I think the macro will also need it.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+                         frag->shm_frag_freed_by_tid);
+         continue;
+      }
+
```
After the `continue`, the body of the for() will just process and do some book-keeping for a single fragment.
And then will `break`.

The unlock is immediately outside the `for()` loop.

I don't there is much to be gained by trying some micro-optimization and run the risk of race conditions creeping in because we unlocked too early.

No change done.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+
+/*
+ * -----------------------------------------------------------------------------
+ * platform_shmcreate() -- Create a new / attach to an existing shared segment.
```
Done.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,136 @@
+platform_shmsize(platform_heap_id heap_id);
+
+size_t
+platform_shmfree(platform_heap_id heap_id);
```
Renamed to `platform_shmbytes_free()` and its sibling to `platform_shmbytes_used()`

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   void *shm_frag_addr;                // Start address of this memory fragment
+                                       // NULL => tracking fragment is empty
+   size_t   shm_frag_size;             // bytes
+   int      shm_frag_allocated_to_pid; // Allocated to this OS-pid
```
`shm_frag_size` is used to check if fragment qualifies based on size-requested during re-allocation.

Rest are for diagnostics, asserts. Added a comment appropriately.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+                         (frag->shm_frag_freed_by_pid ? "free" : "in use"),
+                         frag->shm_frag_freed_by_pid,
+                         frag->shm_frag_freed_by_tid);
+         continue;
```
Not sure it simplifies the code a whole lot. I will have to `break` out of the loop upon finding an element, anyway.

No changes done ...

gapisback commented on src/splinterdb.c:
```diff
@@ -227,38 +229,95 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
+   // has to be calculated here based on other run-time params.
+   // (Some tests externally create the platform_heap, so we should
+   // only create one if it does not already exist.)
+   if (kvs_cfg->use_shmem && (kvs_cfg->heap_handle == NULL)) {
```
@rtjohnso - I reworked this area of handling heap_id. But I did not go as far as making it a separate arg. 

That would mean some more code churn that I wasn't quite ready to take on right now.

gapisback commented on src/task.h:
```diff
@@ -123,6 +123,7 @@ struct task_system {
    // task groups
    task_group group[NUM_TASK_TYPES];
 
+   uint64    ntasks_enqueued;
```
Done as suggested.

gapisback commented on src/task.c:
```diff
@@ -220,8 +221,10 @@ task_invoke_with_hooks(void *func_and_args)
 
-   platform_free(thread_started->ts->heap_id,
-                 thread_started->ts->thread_scratch[thread_started->tid]);
+   void *scratchptr = thread_started->ts->thread_scratch[thread_started->tid];
```
Some test callers, e.g. `test_btree_perf()` call `task_thread_create()` requesting 0 `scratch_size`, so no scratch space would have been allocated.

That's why we have to check for non-NULL `scratchptr` before trying to free that memory.

Added a small comment explaining this usage.

gapisback commented on src/trunk.c:
```diff
@@ -7026,8 +7061,14 @@ trunk_range(trunk_handle  *spl,
+   rc = trunk_range_iterator_init(
       spl, range_itor, start_key, POSITIVE_INFINITY_KEY, num_tuples);
    if (!SUCCESS(rc)) {
       goto destroy_range_itor;
```
@rtjohnso - Usually I'm happy to do this kind of cleanup ... but all this code is quite old and crusty now in my mind.

I don't want to embark on a side-bar code-cleanup fix for an existing code, that is not directly related to the feature development in this change-set. (Sorry ...)

gapisback commented on src/splinterdb.c:
```diff
@@ -379,6 +448,7 @@ splinterdb_close(splinterdb **kvs_in) // IN
    splinterdb *kvs = *kvs_in;
    platform_assert(kvs != NULL);
 
+   splinterdb_close_print_stats(kvs);
```
I have found these stats to be very useful for troubleshooting during test-dev and to diagnose test failures.

Have moved this to only print these stats if shared-memory is configured. That way, normal usage of splinter is unaffected.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -0,0 +1,1152 @@
+   // Optimistically, allocate the requested 'size' bytes of memory.
+   retptr = __sync_fetch_and_add(&shminfo->shm_next, (void *)size);
+
+   if (shminfo->shm_next > shminfo->shm_end) {
```
Thanks, @rtjohnso - yes, subtle bug. Nice find; thank you! Fixed it as suggested.

gapisback commented on src/util.h:
```diff
@@ -142,7 +143,7 @@ writable_buffer_length(const writable_buffer *wb)
 /* May allocate memory */
 platform_status
-writable_buffer_resize(writable_buffer *wb, uint64 newlength);
+writable_buffer_resize(writable_buffer *wb, uint64 oldlength, uint64 newlength);
```
@rtjohnso - We should discuss all these series of comments you have marked-up around the change to `writable_buffer_resize()`.

The way I see it is this: The new arg `oldlength` is only relevant when shared memory is configured. 

It gets passed-down as `writable_buffer_resize()` -> `writable_buffer_ensure_space()` -> `platform_realloc()` -> ... -> `platform_shm_realloc()`.

In the last routine, we use the `oldsize` arg (i.e. old length) to do this:

```
 538       // Copy over old contents, if any, and free that memory piece
 539       if (oldptr) {
 540          memcpy(retptr, oldptr, oldsize);
```

If we don't know the old length, we cannot copy over the required contents.

However, upon looking closer, I realized that we can get the old-length from `wb->length` when calling `platform_realloc()`.

I reworked the interface accordingly, removed the ` oldlength` arg added to `writable_buffer_resize()`.

gapisback commented on src/util.h:
```diff
@@ -229,7 +230,7 @@ writable_buffer_memset(writable_buffer *wb, int c)
 writable_buffer_copy_slice(writable_buffer *wb, slice src)
 {
-   platform_status rc = writable_buffer_resize(wb, slice_length(src));
+   platform_status rc = writable_buffer_resize(wb, 0, slice_length(src));
```
yes, you are right. I have reworked this interface to get at the old-length inside the callee using `wb->length`.

gapisback commented:
HI, @rtjohnso - I finally completed the work to address all (most) of your review comments. And have also rebased v/s latest `/main`.

Here's the overall status of this update cycle:

1. Couple of rebases v/s `/main` went thru smoothly; i.e. there were no conflicts, so I don't think there was any major code churn due to this work. There was one small merge-issue in `trunk.c` which I found and fixed during test cycles.
2. Removed most usages of `platform_heap_handle` from shmem.c and from many external interfaces. There are some vestigial references that need to be cleansed; in future.
3. splinterdb_create() `const` issue has been fixed. New test added to exercise gymnastics.
4. `writable_buffer_resize()` interface change corrected
5. Other minor cleanup and a few bugs that you spotted

Few things I didn't take-on and do ... as they will simply balloon this change-set. And are not super-critical (I think). Some examples:

1. Merging parsing of `--use-shmem` in `config_parse()` and only use that interface to parse this argument.
2. Some other code cleanup ... that I don't recall now.

I expect CI runs to complete ... there IS a lot of new testing going on, so won't be surprised if there are still some failures. I'll watch and stabilize it.

This updated PR is now ready for a final review.
Hopefully we can close out loose threads and try to land this large'ish piece of work.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -617,32 +617,31 @@ platform_shm_track_alloc(shmem_heap *shm, void *addr, size_t size)
-
-      // We found a free slot. Track our memory fragment at fctr'th slot.
+   shm_lock_mem_frags(shm);
+   while ((fctr < ARRAY_SIZE(shm->shm_mem_frags)) && frag->shm_frag_addr) {
```
@rtjohnso -- Redid this as you had asked for ...

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -699,12 +698,16 @@ platform_shm_track_free(shmem_heap *shm,
+   int            fctr = 0;
+   while ((fctr < ARRAY_SIZE(shm->shm_mem_frags))
+          && (!frag->shm_frag_addr || (frag->shm_frag_addr != addr)))
+   {
```
Same change ... Replaced `for()` with `while()` for clarity in code flow ...

gapisback commented on include/splinterdb/splinterdb.h:
```diff
@@ -61,9 +61,6 @@ typedef struct splinterdb_config {
    _Bool  use_shmem; // Default is FALSE.
-   _Bool  trace_shmem_allocs;
-   _Bool  trace_shmem_frees;
-   _Bool  trace_shmem; // Trace both allocs & frees from shared memory
```
Withdrew these. I already had API function to set global Booleans. So, the rework was pretty simple.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -92,12 +92,15 @@ typedef struct shm_frag_info {
+   void *shm_start;      // Points to start address of shared segment.
+   void *shm_end;        // Points to end address; one past end of sh segment
+   void *shm_next;       // Points to next 'free' address to allocate from.
+   void *shm_last_alloc; // Points to address most-recently allocated
```
Used for micro-optimization, under a new spinlock ... See L675

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -106,12 +109,14 @@ typedef struct shmem_heap {
    uint32        shm_num_frags_inuse_HWM;
    size_t        shm_used_by_large_frags; // Actually reserved
    size_t        shm_nfrees;              // # of calls to free memory
+   size_t        shm_nfrees_last_frag;    // Freed last small-fragment
```
New metric ... useful for diags.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -106,12 +109,14 @@ typedef struct shmem_heap {
    size_t shm_used_bytes;  // Used bytes of memory left (that were allocated)
-   uint64 shm_magic;       // Magic identifier for shared memory segment
-   int    shm_id;          // Shared memory ID returned by shmget()
+   size_t shm_used_bytes_HWM; // High-water mark of memory used bytes
```
New metric ... useful for diags.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -585,7 +666,38 @@ platform_shm_free(platform_heap_id hid,
+
+   shm_lock_mem(shm);
+   shm->shm_nfrees++;
+   if (shm->shm_last_alloc == ptr) {
```
If small fragment being freed was the very last one allocated, re-allocate it by grabbing that fragment into 'free space'.

This single optimization pays off hugely for `build/release/bin/unit/splinter_test --use-shmem --num-inserts 20000000 test_lookups`

```
Deallocated SplinterDB shared memory segment at 0x7f3996e51000, shmid=0. Used=3291704704 bytes (~3.06 GiB, 76.64 %) Used HWM=3291704768 bytes (~3.06 GiB, 76.64 %), Free=1003256256 bytes (~956.77 MiB, 23.36 %).
SplinterDB shared memory stats: nfrees=20000706, nfrees-last-small-frag=20000101 (100.00 %), Large fragments in-use HWM=11 (found in-use=0), consumed=3291363008 bytes (~3.06 GiB).
```

Almost all the frees were for the very last fragment allocated; no concurrency in the workload.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -585,7 +666,38 @@ platform_shm_free(platform_heap_id hid,
+   if (maybe_large_frag) {
+      platform_shm_track_free(shm, ptr, objname, func, file, lineno);
+   } else if (0) {
+      platform_default_log(
```
Deleted this dead code; added for pvt debugging.

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -718,6 +829,8 @@ platform_shm_track_free(shmem_heap *shm,
       debug_assert(frag->shm_frag_allocated_to_pid != 0);
       debug_assert(frag->shm_frag_size != 0);
 
+      shm->shm_num_frags_inuse--;
```
Minor accounting bug in older code .... When a large fragment was freed, it was not being account for the 'in-use' counter.

gapisback commented on src/util.c:
```diff
@@ -19,7 +19,8 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
    void *oldptr  = wb->can_free ? wb->buffer : NULL;
-   void *newdata = platform_realloc(wb->heap_id, wb->length, oldptr, minspace);
+   void *newdata = platform_realloc(
+      wb->heap_id, writable_buffer_length(wb), oldptr, minspace);
```
Bug-fix: use the accessor rather than the raw field 'length'.

gapisback commented on test.sh:
```diff
@@ -215,10 +215,10 @@ function nightly_functionality_stress_tests() {
-                                        #       --num-tables ${ntables} \
-                                        #       --cache-capacity-mib ${cache_size} \
-                                        #       --db-location ${dbname}
+    #         "$BINDIR"/driver_test splinter_test --functionality ${num_rows} 1000 \
```
Test-case still remains commented out as it runs into an existing issue.

gapisback commented on test.sh:
```diff
@@ -605,18 +605,19 @@ function run_slower_unit_tests() {
     # Use fewer rows for this case, to keep elapsed times of MSAN runs reasonable.
     msg="Splinter lookups test ${use_msg}"
-    local n_mills=20
+    local n_mills=2
```
Reverted to prev workload; i.e 2 Mil

gapisback commented on src/platform_linux/shmem.c:
```diff
@@ -581,6 +581,16 @@ platform_shm_realloc(platform_heap_id hid,
+                oldsize);
+
+   // We can only realloc from an oldptr that's allocated from shmem
+   debug_assert(!oldptr || platform_valid_addr_in_heap(hid, oldptr),
```
Assertions added just to ensure that no mistakes are made in calling conventions (in future).

gapisback commented on src/util.c:
```diff
@@ -18,17 +18,21 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
-   void *newdata = platform_realloc(
-      wb->heap_id, writable_buffer_length(wb), oldptr, minspace);
+   void *newdata = NULL;
+   if (wb->can_free) {
```
@rtjohnso - Reworked this logic as discussed. Newly added unit-tests seem to work cleanly.

gapisback commented on src/util.c:
```diff
@@ -18,17 +18,21 @@ writable_buffer_ensure_space(writable_buffer *wb, uint64 minspace)
+   void *newdata = NULL;
+   if (wb->can_free) {
+      newdata = platform_realloc(
+         wb->heap_id, wb->buffer_capacity, wb->buffer, minspace);
```
@rtjohnso - Do note there is still a peculiarity here, even though we are using `wb->buffer_capacity` as old-length passed to realloc().

This old-length value (i.e. `wb->buffer_capacity`) will only be used to copy-over data after reallocating new memory.

Your reasoning was that for `free(ptr)`, we should pass-in the allocated memory-size (i.e. buffer-capacity) and not the actual "in-use" data length.

The pair of `(wb->buffer, wb->buffer_capacity)` is not sufficient for `free()` to correctly identify this memory fragment. This is because allocation could have added pad-bytes at the time of allocation, rounding-up requested 'capacity' to something larger (for cache-alignment).

So the argument that `free()` will need to use the `buffer-capacity` rather than `buffer-length` to locate the fragment-to-be-freed is also not kosher.

gapisback commented:
@rtjohnso - I have addressed the last remaining comment that we agreed, during our discussion yesterday, needs to be fixed.

Please see [this commit](https://github.com/vmware/splinterdb/pull/567/commits/997e355b47875abde0e20f214318b3473c3d8dcc), and let me know if this version of the fix is less unacceptable to you.

The new unit-test case added in this commit tripped up in a CI-job, in non-shared-memory execution. I patched an assertion under [this commit](https://github.com/vmware/splinterdb/pull/567/commits/997e355b47875abde0e20f214318b3473c3d8dcc). 

I think it's fine to have this white-box type of unit-test case to clearly understand the gyrations of shared memory allocation. If you strongly object to this tweaking, I can just delete this specific assertion check in the affected test-case, `unit/writable_buffer_test.c`, but that would not be my 1st choice:

```
419    // Currently, reallocation from shared-memory will not reuse the existing
420    // memory fragment, even if there is some room in it for the append. (That's
421    // an optimization which needs additional memory fragment-size info which
422    // is currently not available to the allocator.)
423    if (data->use_shmem) {
424       const void *new_data2 = writable_buffer_data(wb);
425       ASSERT_TRUE((new_data != new_data2));
```

Rest -- ball is in your court. Awaiting to hear of any other changes you'd like to see or other things you'd like to discuss.
Thanks.

rtjohnso reviewed (changes requested):
Here's a review of everything but tests and platform stuff.  A few small changes requested.

gapisback commented on src/clockcache.c:
```diff
@@ -615,6 +615,11 @@ clockcache_get_entry(clockcache *cc, uint32 entry_number)
 }
 
+static inline entry_status
+clockcache_get_flag(clockcache *cc, uint32 entry_number)
```
Done. Renamed as `clockcache_get_status()`.

Most other instances were already accessing using an `entry` ptr, so this accessor function wasn't quite applicable. 

I did find one more candidate, `clockcache_ok_to_writeback() -> clockcache_get_entry()`, which was replaced with a call to this new accessor, `clockcache_get_status()`.

gapisback commented on src/clockcache.c:
```diff
@@ -915,12 +920,12 @@ clockcache_assert_no_locks_held(clockcache *cc)
-   uint64 i;
-
-   for (i = 0; i < cc->cfg->page_capacity; i++) {
+#if SPLINTER_DEBUG
```
Done.

gapisback commented on src/task.c:
```diff
@@ -654,6 +662,7 @@ task_enqueue(task_system *ts,
           > group->stats[tid].max_outstanding_tasks) {
          group->stats[tid].max_outstanding_tasks = group->current_waiting_tasks;
       }
+      group->stats[tid].total_tasks_enqueued += 1;
```
Done. Merged.

gapisback commented on src/splinterdb.c:
```diff
@@ -227,20 +229,52 @@ splinterdb_init_config(const splinterdb_config *kvs_cfg, // IN
+   kvs = TYPED_ZALLOC(use_this_heap_id, kvs);
    if (kvs == NULL) {
       status = STATUS_NO_MEMORY;
+      if (we_created_heap) {
```
Yes, reworked the code to `goto deinit_kvhandle`. Existing end-of-fn exit-handling will take care of required cleanup.

gapisback commented on src/platform_linux/platform_inline.h:
```diff
@@ -415,19 +442,63 @@ platform_aligned_malloc(const platform_heap_id UNUSED_PARAM(heap_id),
     */
-   const size_t padding = (alignment - (size % alignment)) % alignment;
-   return aligned_alloc(alignment, size + padding);
+   const size_t padding  = platform_align_bytes_reqd(alignment, size);
```
I think the way it's coded, it still achieves what we want; i.e., to ensure that the returned 'ptr' is always aligned.

This is the way I argued to myself that this scheme is acceptable.

Let's consider cache-line alignment is 64 bytes.

Say, we did not do this padding up business for every allocation request. Then, after the very 1st allocation of 60 bytes, the `next-addr` will be 60. For the next request, we would have to move the `next-addr` ahead to offset 64, so the returned memory ptr is correctly aligned.

The same thing is being achieved by rounding-up the previous memory request from 60 bytes to 64 bytes.

Essentially, every memory allocation ensures that the `next-addr` ptr is left correctly aligned for the next memory allocation request.

It wastes a tiny bit of space for the very last memory allocation, but for every allocation request before that, I think the net memory usage (wastage) is the same in both schemes.

Hope this makes sense.

rtjohnso replied:
For the non-shmem allocations, this code is unnecessary -- the alignment parameter passed to the lower-level allocator will ensure the requested alignment.

For the shmem allocator, the approach in this code will not always work.  Suppose the `next-addr` starts out 64-byte aligned (which it presumably does).  Suppose the first two allocations performed are:
1. allocate some number of bytes with 32-byte alignment
2. allocate some number of bytes with 64-byte alignment.
The current code will will round up the first allocation to a multiple of 32 bytes, which is fine.  It will leave the `next-addr` to be 32-byte aligned.  But this is not sufficient to ensure that the second allocation gets its desired alignment!  The shmem allocator still needs add some padding _at the beginning of the second allocation_ to ensure that the second allocation request gets its desired alignment.

So, my point is that this code does nothing to actually achieve the alignment of the requested allocations.  There's no reason to round up allocation size requests.  Instead, the underlying shmem allocator itself needs to adjust the pointers it returns to ensure that they are aligned.

gapisback commented on src/platform_linux/shmem.h:
```diff
@@ -0,0 +1,119 @@
+ * Caller-macro to invoke lower-level allocator and to pass-down caller's
+ * context fields, which are printed for diagnostics under a traceflag.
+ */
+#define splinter_shm_alloc(heap_id, nbytes, objname)                           \
```
Removed all 3 caller macros, and changed code to directly invoke one of `platform_shm_alloc()`, `platform_shm_free()` or `platform_shm_realloc()`.

gapisback commented on src/platform_linux/platform_types.h:
```diff
@@ -161,7 +161,7 @@ typedef struct {
 
 typedef void *platform_module_id;
-typedef void *platform_heap_handle;
+typedef char *platform_heap_handle;
```
This change was done to ferret out bad callers (mainly from shmem.c) who might be inadvertently calling some function with `platform_heap_handle`, while they were supposed to pass-in `platform_heap_id hid`.

This change was helpful to track down such errant callers.

Now that callers to `platform_heap_handle` have been eliminated from the code, L164 can also be deleted.

Cleaned this up ... We will no longer see references to `platform_heap_handle` in the code base.

gapisback commented on tests/config.c:
```diff
@@ -308,13 +332,45 @@ config_parse(master_config *cfg, const uint8 num_config, int argc, char *argv[])
          // Test-execution configuration parameters
          config_set_uint64("seed", cfg, seed) {}
          config_set_uint64("num-inserts", cfg, num_inserts) {}
+         config_set_uint64("num-threads", cfg, num_threads) {}
```
@rtjohnso  - I didn't do anything for the following reason. (I should have left a note previously.)

The existing functional `splinter_test.c` already supports `--num-insert-threads`, ` --num-lookup-threads`,  `--num-range-lookup-threads`,  and `--num-pthreads`. However, the parsing for these flags is done in the hand-rolled parser: `splinter_test_parse_perf_args()`.

So ... I didn't see any need to break THIS up into number of query threads argument.

This new argument, `num-threads` is only being used in the new `large_inserts_bugs_stress_test.c`, whose fate is still unclear to me.

Currently all calls to that test in `test.sh` are commented out.

I should probably just delete L367 above ... but, first, wanted to hear what you'd like to be done with this new test.

gapisback replied:
We agreed to delete the newly added `--num-threads` for now.

gapisback commented:
@rtjohnso - I've addressed your last batch of comments that needed code changes. 

Only a couple to do with config-parsing were not addressed --I've left my reasons in the responses at the appropriate file.

The one thing remaining is: What would you like the random value to define `SPLINTERDB_SHMEM_MAGIC` to be?

I could not think of anything better than what's already in the code. In any case, it's only for diagnostics, so does it really matter that the value defined be random, more than what it's currently defined to?

gapisback commented:
**Summary comments**:

- If `--use-shmem` is enabled in master_cfg, ensure that `splinterdb_create()` does create shared memory.
- Evaluate what more / other tests are enabled, or can be enabled, with this approach. Include all tests that can be run with this new interface as part of `test.sh`.
- Figure out a way to engage the new `large_inserts_bugs_stress_test` tests thru `test.sh`, but make those failures non-blocking to CI. Handle failures in `test.sh` so the execution does not fail immediately.

gapisback commented:
HI, @rtjohnso -- I investigated the issues around padding for cache-alignment that you raised in [this review comment](https://github.com/vmware/splinterdb/pull/567/files#r1295148145) and in [this comment](https://github.com/vmware/splinterdb/pull/567/files#r1295370859).

The affected code chunk in `/main` is this one, and has been in the code line since [the very initial commit](https://github.com/vmware/splinterdb/commit/f94168143) created to move this code base into GitHub.

```
403 static inline void *
404 platform_aligned_malloc(const platform_heap_id UNUSED_PARAM(heap_id),
405                         const size_t           alignment, // IN
418    const size_t padding = (alignment - (size % alignment)) % alignment;
419    return aligned_alloc(alignment, size + padding);
```

I looked at the semantics of `aligned_alloc()` and here is what I found in Gnu docs [Aligned-Memory-Blocks.html](https://www.gnu.org/software/libc/manual/html_node/Aligned-Memory-Blocks.html):

```
The aligned_alloc function allocates a block of size bytes whose address is a multiple of alignment. The alignment must be a power of two and size must be a multiple of alignment. 
```
[Linux man page](https://linux.die.net/man/3/aligned_alloc) also documents something similar:

```
The function aligned_alloc() is the same as memalign(), except for the added restriction that size should be a multiple of alignment. 
```

I think this answers a question you raised in a previous comment; i.e. "_Alignment is a request on the position of the start of the returned allocation. Why would we need to pad the the size?_"

Given that this is the expected behaviour of `aligned_alloc()`, do you still see the need to rework shared-memory based allocator, which currently also inherits this padding-up of requested `size` to `required` bytes?

I think it's a reasonable choice to keep both the schemes symmetric in this regard.

---
In any case, I did try to actually re-work `platform_shm_alloc()` to receive an `alignment` argument and then to do the alignment of `shm->shm_next` inside `platform_shm_alloc()`. It will need a bit of pointer-arithmetic / bit-masking to get this logic right.

But before embarking on finalizing that code rework, I wanted to bring this semantics issue vis-a-vis `aligned_alloc()` to see you have a re-think.

As mentioned earlier, having to rework this now will make it a bit more tricky to merge with the upcoming `memfrag{}`-based memory allocation / free interfaces (in part-3).

If you still do think that this padding business must be reworked for shared-memory based allocation, let me suggest that we revisit that once part-3 lands, at which time this whole memory alloc / free picture will become clearer.

--

Let me know your thoughts based on this new information.

gapisback commented:
HI, @rtjohnso -- Barring the one remaining issue of how-to handle the business of requests for aligned memory (see [previous response](https://github.com/vmware/splinterdb/pull/567#issuecomment-1684880355) of mine), here is the status of the remaining rework based on our discussion this past week.

All items you raised as needing changes have been addressed, with either code-rework, test-changes or extended enablement of test executions via `test.sh`.

Main changes arising in last round of updates, to address your review comments summarized in [this entry](https://github.com/vmware/splinterdb/pull/567#issuecomment-1681240300):

- Withdrew the special-case test_using_shmem() parsing interface with an equivalent interface,
  config_parse_use_shmem(), which uses existing config_parse() to parse the --use-shmem
  argument. Applied this change to all functional and unit-tests, where applicable.
  Now, we no longer have the restriction that --use-shmem must be the very 1st argument
  while invoking tests to run with shared memory configured.

- Reworked clockcache_assert_clean() to become a boolean routine called under debug_assert().

- Re-enabled few cases of large_inserts_bugs_stress_test unit-tests in test.sh . As these
  are known to sporadically fail, bracketted the call under set +e/-e to avoid percolating
  failures to the entire test execution. (Need to watch CI test-runs to see if this fix
  will allow CI jobs to 'pass' inspite of local failures.)

- Extended coverage in test.sh [1] and [2] to run all the existing tests in `--use-shmem` mode

- Other minor code cleanup.

`test.sh` has a new driver function, `run_tests_with_shared_memory()`, which calls these sub-tests executed with `--use-shmem` option.

[1] **** Unit Tests ****: `unit_test "--use-shmem"`, and `run_slower_unit_tests "--use-shmem"` invokes these unit-tests.


cc-sdb-vm:[69] $ build/release/bin/unit_test --list

List of test suites that can be run with shared-memory configured:
```
  task_system                   - Yes
  splinterdb_stress             - Yes
  splinterdb_quick              - Yes
  splinterdb_heap_id_mgmt       - Yes
  splinter_shmem                - Yes
  limitations                   - Yes
  btree                         - Yes
  btree_stress                  - Yes
  writable_buffer               - Yes
  util                          - No
  platform_api                  - No
  misc                          - No
  splinter                      - Yes

 - unit/config_parse_test.c : Not relevant to feature
 - unit/misc_test.c : Not relevant to feature
 - unit/platform_apis_test.c: Does not add value to convert this test;
 - unit/util_test.c: Not relevant to feature
                               platform_heap_create() with use_shmem is
                               tested in many other tests.
```

[2] **** Functional tests that can be run with shared-memory configured ****

See calls to run_splinter_perf_tests(), run_btree_tests() and run_other_driver_tests() in `test.sh`

```
 - splinter_test : Yes; Both --functionality and --perf
 - btree_test    : Yes; Both basic and --perf
 - cache_test   : Yes
 - log_test      : Yes
 - filter_test   : Yes
```

rtjohnso reviewed (approved):
Looks good to go!
